### ISSUE TITLE
Rule proposal: `no-barrel-files` or `no-re-export`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#3520]
### Description

In general, importing and re-exporting a value only *adds* a way to import it since the original export remains accessible. 

On this basis, `export-from` should never be used on files that share the same visibility. 

The `exports` field in package.json is what changes the "visibility". 

### Examples

```js
// ❌
export {value} from './file.js'

// ❌
import {value} from './file.js'
export {value}

// ✅ import directly from the original file *where needed*
import {value} from './file.js'
```

### Proposed rule name

no-re-export

### Additional Info

To keep things simpler (without having to determine the "visibility" of any given file) and to follow the core principle, the rule could be implemented *regardless* of visibility, even if this means pushing libraries away from barrel files altogether. 

```js
// ❌
export {JsonValue} from './JsonValue.js'


// ✅ import directly from the original file *where needed*
import {JsonValue} from 'type-fest/json-value'
```

`type-fest` would actually be greatly affected by this, both because it would turn one import into many and because TS will no longer need to parse *the entire library* when you import a single type. 

This article contains more details: https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/

**Comment by fregante** [comment]
The rule could be implemented with an option:

- `allowExportsFromInternalFiles`: when true, it lets type-fest exist as is

**Comment by fregante** [comment]
I'm not sure if exporting values from dependencies should be allowed. For example:

```js 
export {SerializedError} from 'serialize-error'
export default function myModule() {
  // does something with serialize-error
}
```
This way users of myModule don't have to directly depend on `serialize-error` to use myModule's types.

**Comment by sindresorhus** [comment]
The broad `no-re-export` proposal would require package visibility and a module graph, which is not feasiable, so I think the useful rule should be narrower: `no-useless-re-export`.

It reports a direct named re-export only when the same module is already re-exported with `export *`:

```js
export * from './foo.js';
export {foo} from './foo.js'; // reported
```

**Comment by fregante** [comment]
That's not what I asked though 🥺 

The core issue is that barrel files and re-exports are bad for performance and duplication. I'd say that if you do want barrel files, then you should disable the `no-exports-from` rule. 

Packages like type-fest, if they want to keep offering bad performance, can have a single `eslint-disable` in their package entry point.

**Comment by sindresorhus** [comment]
Ah ok. I misunderstood. I agree this could be useful as a project-specific rule, but I also don't think it's a good fit here. To work reliably it would require knowing the module graph, so probably a better fit for https://github.com/un-ts/eslint-plugin-import-x ?

**Comment by fregante** [comment]
The rule can exist in many different strictness levels, but as far as I'm concerned, `export from` is never acceptable, `ExportNamedDeclaration[source]` alone should cover this. 

Exceptions you might like are "one barrel file per package", like type-fest does, but you don't need the module graph for this. To me it's easier to mark the single file via `/* eslint-disable no-re-export -- Entry point */` than letting anyone adding random `export {SomeTime} from './SomeType.js'` anywhere.

**Comment by fregante** [comment]
Case in point:

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/af8fd60b3a05e8fdbdfe5af17404397fc5689698/rules/utils/index.js#L1-L157

You (or Claude) are maintaining this massive barrel file that does nothing. Any contributor can still import the same values from the original files.

**Comment by sindresorhus** [comment]
I think this is a better fit for `eslint-plugin-import-x`. A useful rule needs module resolution and project-wide context to distinguish public API entry points from problematic barrels. import-x already owns resolver-aware import/export analysis; a syntax-only Unicorn rule would produce false positives and overlap with `no-useless-re-export` and `prefer-export-from`.

**Comment by fregante** [comment]
> distinguish public API entry points from problematic barrels

All barrels are problematic IMHO, that's the point. That's why resolution is not a limiting factor here.

https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/
